src/Chainweb/OpenAPIValidation.hs

```diff
@@ -0,0 +1,70 @@
+{-# language TupleSections #-}
+{-# language ScopedTypeVariables #-}
+{-# language FlexibleContexts #-}
+{-# language OverloadedStrings #-}
+
+module Chainweb.OpenAPIValidation (mkValidationMiddleware) where
+
+import Control.Monad
+import Data.Aeson
+import qualified Data.ByteString.Char8 as BS8
+import qualified Data.ByteString.Lazy as BL
+import Data.Foldable
+import qualified Data.HashSet as HS
+import Data.IORef
+import Data.LogMessage (LogFunctionText)
+import qualified Data.Text.Encoding as T
+import qualified Data.Yaml as Yaml
+import qualified Network.HTTP.Client as HTTP
+import Network.Wai
+import qualified Network.Wai.Middleware.Validation as WV
+import System.LogLevel
+
+import Chainweb.Logger
+import Chainweb.Time
+import Chainweb.Utils
+import Chainweb.Version
+
+mkValidationMiddleware :: Logger logger => logger -> ChainwebVersion -> HTTP.Manager -> IO Middleware
+mkValidationMiddleware logger v mgr = do
+    (chainwebSpec, pactSpec) <- fetchOpenApiSpecs
+    apiCoverageRef <- newIORef $ WV.initialCoverageMap [chainwebSpec, pactSpec]
+    apiCoverageLogTimeRef <- newIORef =<< getCurrentTimeIntegral
+    return $
+        WV.mkValidator apiCoverageRef (WV.Log logValidationFailure (logApiCoverage apiCoverageLogTimeRef)) $ \path -> asum
+            [ case BS8.split '/' path of
+                ("" : "chainweb" : "0.0" : rawVersion : "chain" : rawChainId : "pact" : "api" : "v1" : rest) -> do
+                    findPact pactSpec rawVersion rawChainId rest
+                ("" : "chainweb" : "0.0" : rawVersion : "chain" : rawChainId : "pact" : rest) -> do
+                    findPact pactSpec rawVersion rawChainId rest
+                _ -> Nothing
+            , (,chainwebSpec) <$> BS8.stripPrefix (T.encodeUtf8 $ "/chainweb/0.0/" <> chainwebVersionToText v) path
+            , Just (path,chainwebSpec)
+            ]
```

Comment:
Why do we have to do this kind of analysis for determining which endpoint the request has came to? Why not make two independent validator middlewares, one for the p2p and another for the service endpoints and inject them separately into the corresponding servers?

Reply:
> Why not make two independent validator middlewares, one for the p2p and another for the service endpoints and inject them separately into the corresponding servers?
That isn't quite what this does. We have two OpenAPI specs, one for the Pact API, and one for the chainweb-node API. The service API includes both the chainweb-node API and the Pact API, the P2P API includes only part of the chainweb-node API. So while the Pact stuff could be excluded from the P2P API, this code would still be necessary for the service API endpoint at least.

Reply:
I see, that makes sense, thanks! 

There's some non-trivial coupling here. As far as I can tell, this mixes the routing of the request to the correct API spec with some pre-processing of the `path` before it's passed to the validator. Why do we need this kind of path pre-processing? I guess this means there's a gap between the API spec and the actual API, so where does this gap come from, is it something like a middleware defined by the `chainweb-node` codebase? Or does it come from the behavior of some external middleware like `nginx`? Is this special path pre-processing explained inside the spec itself, or maybe it's supposed to be legacy?  

Anyway, I guess these questions are mostly outside the scope of this PR, but I think it could be nice to decouple the spec-selection from the spec-specific path pre-processing. For example how about a refactor like this to make it more clear: https://github.com/kadena-io/chainweb-node/commit/0522c11c25ace2eaa0dc435499e30b4d7ecc30ba I usually do this kind of decoupling so that I have natural places in the code to explain why "surprising" logic like this exists, maybe write a comment to point to another place in the code that we're implementing the inverse of?